src/aggregation/bucket/filter.rs

```diff
@@ -0,0 +1,1922 @@
+        });
+
+        assert_agg_results!(&result, expected, 1.0);
+        Ok(())
+    }
+
+    // ============================================================================
+    // Nested Filter Tests
+    // ============================================================================
+
+    #[test]
+    fn test_two_level_nested_filters() -> crate::Result<()> {
+        let index = create_standard_test_index()?;
+        let reader = index.reader()?;
+        let searcher = reader.searcher();
+
+        let agg = json!({
+            "all": {
+                "filter": "*",
+                "aggs": {
+                    "electronics": {
+                        "filter": "category:electronics",
+                        "aggs": {
+                            "expensive": {
+                                "filter": "price:[900 TO *]",
+                                "aggs": {
+                                    "count": { "value_count": { "field": "brand" } }
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+        });
+
+        let aggregations: Aggregations = serde_json::from_value(agg)?;
+        let collector = create_collector(&index, aggregations);
+        let result = searcher.search(&AllQuery, &collector)?;
+
+        let expected = json!({
+            "all": {
+                "doc_count": 4,
+                "electronics": {
+                    "doc_count": 2,
+                    "expensive": {
+                        "doc_count": 1,  // Only apple (999) is >= 900
+                        "count": { "value": 1.0 }
+                    }
+                }
+            }
+        });
+
+        assert_agg_results!(&result, expected);
+        Ok(())
+    }
+
+    #[test]
+    fn test_deeply_nested_filters() -> crate::Result<()> {
+        let index = create_standard_test_index()?;
+        let reader = index.reader()?;
+        let searcher = reader.searcher();
+
+        let agg = json!({
+            "level1": {
+                "filter": "*",
+                "aggs": {
+                    "level2": {
+                        "filter": "in_stock:true",
+                        "aggs": {
+                            "level3": {
+                                "filter": "rating:[4.0 TO *]",
+                                "aggs": {
+                                    "level4": {
+                                        "filter": "price:[500 TO *]",
+                                        "aggs": {
+                                            "final_count": { "value_count": { "field": "brand" } }
+                                        }
+                                    }
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+        });
+
+        let aggregations: Aggregations = serde_json::from_value(agg)?;
```

Comment:
All the tests should include a deser roundtrip
```suggestion
        let aggregations: Aggregations = serde_json::from_str(& serde_json::to_string(serde_json::from_value(agg)?).unwrap()).unwrap();
```

Reply:
Done.